pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package abcupdater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/abc-updater/pkg/abcupdater/localstore"
+	"github.com/abcxyz/pkg/logging"
+)
+
+const (
+	installIDFileName = "id.json"
+)
+
+var regExUUID = regexp.MustCompile("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
+
+type metricsConfig struct {
+	ServerURL string `env:"ABC_UPDATER_METRICS_URL,default=https://abc-updater-metrics.tycho.joonix.net"`
+}
+
+// MetricsInfo are the parameters for sending metrics.
+type MetricsInfo struct {
+	// The ID of the application to check.
+	AppID string
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string
+
+	// An optional Lookuper to load envconfig structs. Will default to os environment variables.
+	Lookuper envconfig.Lookuper
+
+	Metrics map[string]int
```

Comment:
There are a lot of design decisions contained in the choice of type here.

 - Maybe int64 instead of in case we want to talk about large numbers?
 - Will we ever want floating point metrics like...  um... maybe a ratio? Dunno
 - Do we ever want to collect events, or just numbers ("a $THING occurred")? Would that be represented as a count of 0?
 - Do we want to collect errors?
 - Do we want to collect metrics that are themselves maps? Like "I got error X while rendering template Y"

Did we consider using something like OpenCensus? If not, we could still be inspired by their API, which is a more complicated alternative to just using a map[string]int (Go docs [here](https://opencensus.io/quickstart/go/metrics/)).

This might be one of those rare cases where even I think a design doc might be helpful.

Reply:
I'll send you the design doc. Its partially inherited so I don't know all the thinking around why we didn't just use an existing system.

Reply:
There have been a few design docs, but the goal is to produce usage metrics, not implement something like client-side Sentry. We're primarily interested in the number of unique daily active users.